File: LLD/Design-Patterns/Decorator/code/Decorator.cpp
// Decorator Pattern — coffee shop add-ons, stacked at runtime.
// Compile: g++ -std=c++14 Decorator.cpp -o decorator
// Run:     ./decorator

#include <iostream>
#include <string>
#include <memory>

class Coffee {
public:
    virtual double getCost() const = 0;
    virtual std::string getDescription() const = 0;
    virtual ~Coffee() = default;
};

class Espresso : public Coffee {
public:
    double getCost() const override { return 2.00; }
    std::string getDescription() const override { return "Espresso"; }
};

class CoffeeDecorator : public Coffee {
public:
    explicit CoffeeDecorator(std::unique_ptr<Coffee> wrapped) : wrapped_(std::move(wrapped)) {}

protected:
    std::unique_ptr<Coffee> wrapped_;
};

class Milk : public CoffeeDecorator {
public:
    explicit Milk(std::unique_ptr<Coffee> wrapped) : CoffeeDecorator(std::move(wrapped)) {}
    double getCost() const override { return wrapped_->getCost() + 0.50; }
    std::string getDescription() const override { return wrapped_->getDescription() + " + Milk"; }
};

class Caramel : public CoffeeDecorator {
public:
    explicit Caramel(std::unique_ptr<Coffee> wrapped) : CoffeeDecorator(std::move(wrapped)) {}
    double getCost() const override { return wrapped_->getCost() + 0.75; }
    std::string getDescription() const override { return wrapped_->getDescription() + " + Caramel"; }
};

int main() {
    std::unique_ptr<Coffee> plain = std::make_unique<Espresso>();
    std::cout << plain->getDescription() << " = $" << plain->getCost() << std::endl;

    std::unique_ptr<Coffee> fancy = std::make_unique<Caramel>(std::make_unique<Milk>(std::make_unique<Espresso>()));
    std::cout << fancy->getDescription() << " = $" << fancy->getCost() << std::endl;

    return 0;
}

File: LLD/Design-Patterns/Decorator/code/Decorator.java
// Decorator Pattern — coffee shop add-ons, stacked at runtime.
// Compile: javac Decorator.java
// Run:     java Decorator

interface Coffee {
    double getCost();
    String getDescription();
}

class Espresso implements Coffee {
    public double getCost() { return 2.00; }
    public String getDescription() { return "Espresso"; }
}

abstract class CoffeeDecorator implements Coffee {
    protected Coffee wrapped;

    CoffeeDecorator(Coffee wrapped) {
        this.wrapped = wrapped;
    }
}

class Milk extends CoffeeDecorator {
    Milk(Coffee wrapped) { super(wrapped); }
    public double getCost() { return wrapped.getCost() + 0.50; }
    public String getDescription() { return wrapped.getDescription() + " + Milk"; }
}

class Caramel extends CoffeeDecorator {
    Caramel(Coffee wrapped) { super(wrapped); }
    public double getCost() { return wrapped.getCost() + 0.75; }
    public String getDescription() { return wrapped.getDescription() + " + Caramel"; }
}

public class Decorator {
    public static void main(String[] args) {
        Coffee plain = new Espresso();
        System.out.println(plain.getDescription() + " = $" + plain.getCost());

        Coffee fancy = new Caramel(new Milk(new Espresso()));
        System.out.println(fancy.getDescription() + " = $" + fancy.getCost());
    }
}
